### PR TITLE
Explicit phase offset as a fittable parameter

### PR DESCRIPTION
**Pull request by abhisrkckl**
This PR implements a `PhaseOffset` component that applies an overall phase offset (`PHOFF`) between physical TOAs and the TZR TOA. This works by using a new attribute `tzr` in the `TOAs` class that distinguishes between physical and TZR TOAs.

The sign of `PHOFF` is set to be the same as that of the implicit `Offset`. The design matrix entries for `PHOFF` and `Offset` should be identical.

If the `PhaseOffset` component is present in the timing model, implicit offset (`Offset`) will not be included in the design matrix and the correlation matrix, and the mean will not be subtracted from the residuals.

See issues #38 and #431 and PR #443. 
#1550 was an attempt to do this that didn't work.

### CODE REVIEW

**Comment by abhisrkckl** [comment]
I think this works and is ready to go.

**Comment by dlakaplan** [comment]
Can you give some examples of how this would work?  Have you added discussion to the docs?

**Comment by dlakaplan** [comment]
Can you share the links to the RTD sections you updated?

**Comment by dlakaplan** [comment]
That's good.  Maybe also include a snippet showing how to add just a parameter?

**Comment by abhisrkckl** [comment]
I am working on an example.

File: src/pint/residuals.py [hunk 1]
```diff
@@ -107,7 +108,7 @@ def __init__(
         self.toas = toas
         self.model = model
         self.residual_type = residual_type
-        self.subtract_mean = subtract_mean
+        self.subtract_mean = subtract_mean and "PhaseOffset" not in model.components
```
Comment (dlakaplan) [inline src/pint/residuals.py:111]: Maybe a `debug` message when the presence of a PhaseOffset overrides `subtract_mean`?

File: src/pint/toa.py [hunk 2]
```diff
@@ -1255,6 +1255,8 @@ class TOAs:
         The TOA objects this TOAs should contain.
     toatable : astropy.table.Table, optional
         An existing TOA table
+    tzr : bool
+        Whether TZR TOA
```
Comment (dlakaplan) [inline src/pint/toa.py:1259]: seems like this is missing some text

File: src/pint/toa.py [hunk 3]
```diff
@@ -1291,9 +1293,11 @@ class TOAs:
         available to use names as compatible with TEMPO as possible.
     wideband : bool
         Whether the TOAs also have wideband DM information
+    tzr : bool
+        Whether TZR TOA
```
Comment (dlakaplan) [inline src/pint/toa.py:1297]: same.

**Comment by dlakaplan** [comment]
I think the tests should:
* show what happens if you try to fit for both TZRMJD and PHOFF
* show how the residuals change when you change PHOFF
* show that the residuals are not mean subtracted

**Comment by paulray** [comment]
> * show what happens if you try to fit for both TZRMJD and PHOFF

I didn't think TZRMJD was ever a fittable parameter. Is it?

**Comment by paulray** [comment]
Maybe in the doc paragraph you can say that once you have a non-zero PHOFF, you can make it zero by computing d_phase/d_TOA at the time of the TZR TOA and turning that into a time difference required to zero out the PHOFF?  I assume that is how TEMPO and Tempo2 output their updated TZR TOAs after a fit?

File: src/pint/toa.py (hunk 3 above)
Comment (abhisrkckl) [reply src/pint/toa.py]: fixed

File: src/pint/toa.py (hunk 2 above)
Comment (abhisrkckl) [reply src/pint/toa.py]: fixed

File: src/pint/residuals.py (hunk 1 above)
Comment (abhisrkckl) [reply src/pint/residuals.py:116]: added debug message

**Comment by abhisrkckl** [comment]
@dlakaplan Added the example `phase_offset_example.py`

**Comment by dlakaplan** [comment]
The example looks great.  Probably that's sufficient.  I don't know if we also want a few lines in the docstring for `PhaseOffset` showing how to add one. Maybe just a link to the example?

**Comment by abhisrkckl** [comment]
The process of adding the explicit offset is pretty much the same as any other component/parameter. i.e., add `PHOFF` to the par file or add `PhaseOffset` to the model using the `add_component` method. Do we need separate instructions for that?

**Comment by dlakaplan** [comment]
Probably not.  so maybe just a link to the example in the docstring.

**Comment by dlakaplan** [comment]
OK, great.  I can merge as soon as it passes (note it seems like there are GitHub errors)

**Pull request merged**